### ISSUE TITLE
Doctrine Embeddable Support

### ISSUE THREAD

**Issue by bobdercole** [K-Phoen/rulerz#26]
Hello,

First off, thank you for this awesome library! I am attempting to utilize RulerZ with Doctrine. I have a basic working implementation thanks to the [cookbook](https://github.com/K-Phoen/rulerz/blob/master/doc/cookbooks/doctrine_orm.md). However, I am facing an issue with [Doctrine Embeddables](http://doctrine-orm.readthedocs.org/projects/doctrine-orm/en/latest/tutorials/embeddables.html).

Here is a snippet of my entity map file:

``` yaml
Website:
    type: entity
    table: websites
    embedded:
        domain_name:
            class: DomainName
```

Here is a snippet of my embeddable map file:

``` yaml
DomainName:
  type: embeddable
  fields:
    value:
        type: string
        nullable: false
        length: 255
        options:
            fixed: false
        column: url
```

Here is a snippet of my RulerZ specification file:

``` php
public function getRule()
{
    return 'domain_name.value = :domain_name';
}

public function getParameters()
{
    return [
        'domain_name' => $this->domain_name
    ];
}
```

According to the [Doctrine documentation](http://doctrine-orm.readthedocs.org/projects/doctrine-orm/en/latest/tutorials/embeddables.html#dql), the DQL should look like this:

``` sql
SELECT w FROM Website w WHERE w.domain_name.value = :domain_name
```

However, RulerZ is treating the embeddable as an association. The generated DQL looks like this:

``` sql
SELECT w FROM Website w INNER JOIN w.domain_name rulerz_domain_name WHERE rulerz_domain_name.value = :domain_name
```

What am I doing wrong?

**Comment by K-Phoen** [comment]
Hi!

You're not doing anything wrong, there is just no embeddable support in RulerZ at the moment. That's something I want to see in this library but unfortunately I didn't have the time to implement it yet.

**Comment by bobdercole** [comment]
I looked into how we could achieve embeddable support. Here is my proposal:
- Modify RulerZ\Executor\DoctrineQueryBuilder\AutoJoin to analyse all of the Doctrine class meta data involved in the query.
- The meta data analysis would need to check for embeddables in the root entities and all of their associations.
- The list of embeddables could then be used to decide the outcome of the auto-join.

I believe this is a decent solution because it keeps entity association/embeddable implementations abstracted from the specification. One potential downside is that the AutoJoin class may need to recursively call EntityManager's getClassMetadata method to find all of the embeddables. I understand that this may introduce a negative performance impact.

Let me know if this makes sense to you or if you have any suggestions. I do have a working prototype if you would like to see it.

**Comment by K-Phoen** [comment]
I also had this idea.

Feel free to open a PR, I'd be happy to review it :)

**Comment by K-Phoen** [comment]
By the way, do not hesitate to open a PR even if it's still a WIP ;)

**Comment by bobdercole** [comment]
Sure. Here it is: #28

**Comment by K-Phoen** [comment]
Fixed by #28
